File: scripts/exec_model/layout.py
"""Partition layout vocabulary — the properties a node declares about its output.

Declarations only; no execution logic. Mirrors the Rust types in
`llm-wiki/architecture.md` (Execution / Traits).
"""

from __future__ import annotations

from dataclasses import dataclass, field
from enum import Enum


class NodeKind(Enum):
    SOURCE = "Source"
    INTERMEDIATE = "Intermediate"
    SINK = "Sink"


class BatchLayout(Enum):
    SINGLE_BATCH = "SingleBatch"
    MULTIPLE_BATCHES = "MultipleBatches"


@dataclass(frozen=True)
class ColumnOrder:
    column: int
    ascending: bool = True
    nulls_first: bool = False


class KeyDistributionKind(Enum):
    NOT_SPECIFIED = "NotSpecified"
    BY_HASH = "ByHash"


@dataclass(frozen=True)
class KeyDistribution:
    """Spark murmur3, seed 42 — the only routing `GpuEmitPartitions` has."""

    kind: KeyDistributionKind = KeyDistributionKind.NOT_SPECIFIED
    hash_keys: tuple[int, ...] = ()

    @classmethod
    def not_specified(cls) -> "KeyDistribution":
        return cls()

    @classmethod
    def by_hash(cls, hash_keys) -> "KeyDistribution":
        return cls(KeyDistributionKind.BY_HASH, tuple(hash_keys))

    def is_subset_of(self, group_columns) -> bool:
        """The `hashKeys ⊆ group columns` rule a final aggregate's input must satisfy."""
        if self.kind is KeyDistributionKind.NOT_SPECIFIED:
            return False
        return set(self.hash_keys).issubset(set(group_columns))


class SortKind(Enum):
    """Two-valued on purpose.

    A third `PartitionSorted` — whole stream ordered, not merely each batch — would be
    a synonym: the only nodes that order a whole stream (`GpuAccumulateBatchesAndSort`,
    `GpuMergeSortedPartitions`) emit exactly one batch, so the property holds precisely
    when `BatchSorted` meets `SingleBatch` and `PartitionLayout.is_stream_sorted` derives
    it. It becomes a real third state only under #138's ranged merge emission, which
    orders a stream across several batches.
    """

    NOT_SPECIFIED = "NotSpecified"
    BATCH_SORTED = "BatchSorted"


@dataclass(frozen=True)
class SortOrder:
    kind: SortKind = SortKind.NOT_SPECIFIED
    columns: tuple[ColumnOrder, ...] = ()

    @classmethod
    def not_specified(cls) -> "SortOrder":
        return cls()

    @classmethod
    def batch_sorted(cls, columns) -> "SortOrder":
        return cls(SortKind.BATCH_SORTED, tuple(columns))

    @property
    def is_batch_sorted(self) -> bool:
        return self.kind is SortKind.BATCH_SORTED


class UniqueScope(Enum):
    """How far a uniqueness guarantee reaches."""

    PER_BATCH = "PerBatch"          # no two rows of one batch share the key
    PER_PARTITION = "PerPartition"  # …of one lane, across its batches
    GLOBAL = "Global"               # …of the whole output


@dataclass(frozen=True)
class UniqueKeys:
    """A key set this output is unique on, and how widely.

    Declared and *not* checked. An aggregate knows the scope for free — its init is unique
    per batch, its per-lane merge per partition, and a final under a shuffle on those keys
    globally — and later work can read it rather than re-deriving it: a distinct over an
    already-globally-unique key is a no-op, a join on a globally unique build key cannot
    fan out, and a merge whose input is unique per partition needs no re-aggregation.
    Recording it now costs nothing and means the fact is not lost by the node that knew it.
    """

    columns: tuple[int, ...]
    scope: UniqueScope


@dataclass(frozen=True)
class PartitionLayout:
    n: int
    key_distribution: KeyDistribution = field(default_factory=KeyDistribution.not_specified)
    sort_order: SortOrder = field(default_factory=SortOrder.not_specified)
    batch_layout: BatchLayout = BatchLayout.MULTIPLE_BATCHES
    #: key sets this output is unique on; declared for later use, never checked here
    unique_keys: tuple[UniqueKeys, ...] = ()

    @property
    def is_stream_sorted(self) -> bool:
        """Whole stream ordered, not merely each batch — what a top-N after a sort needs.

        Derived rather than declared, so there is no second way to say it and no pair of
        fields that can disagree.
        """
        return self.sort_order.is_batch_sorted and self.batch_layout is BatchLayout.SINGLE_BATCH

File: peacockdb-core/src/batch_partitioned/layout.rs
//! What a node declares about its output: how many lanes, how rows were routed into
//! them, what order they carry and whether a lane is one batch. Declarations only —
//! nothing here executes, and the vocabulary is fixed by `llm-wiki/architecture.md`.

use super::schema::Schema;

/// A sink structurally has no layout and no schema; everything else always has both,
/// which is why they live inside the kind rather than beside it as two `Option`s that
/// have to be `None` together.
#[derive(Debug, Clone, PartialEq)]
pub enum NodeKind {
    Source {
        layout: PartitionLayout,
        schema: Schema,
    },
    Intermediate {
        layout: PartitionLayout,
        schema: Schema,
    },
    Sink,
}

impl NodeKind {
    /// `None` for a sink, which structurally has neither.
    pub fn layout(&self) -> Option<&PartitionLayout> {
        match self {
            Self::Source { layout, .. } | Self::Intermediate { layout, .. } => Some(layout),
            Self::Sink => None,
        }
    }

    pub fn schema(&self) -> Option<&Schema> {
        match self {
            Self::Source { schema, .. } | Self::Intermediate { schema, .. } => Some(schema),
            Self::Sink => None,
        }
    }
}

/// One sort key: a column ordinal into the declaring node's schema, and its direction.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct ColumnOrder {
    pub column: u32,
    pub ascending: bool,
    pub nulls_first: bool,
}

/// How rows were routed into lanes. `ByHash` is Spark murmur3 seed 42 — the only
/// routing `GpuEmitPartitions` has.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum KeyDistribution {
    NotSpecified,
    ByHash { hash_keys: Vec<u32> },
}

impl KeyDistribution {
    /// The `hash_keys ⊆ group columns` rule a final aggregate's input must satisfy:
    /// rows of one group are co-located only when the shuffle keyed on a subset of
    /// the columns being grouped.
    pub fn is_subset_of(&self, group_columns: &[u32]) -> bool {
        match self {
            Self::NotSpecified => false,
            Self::ByHash { hash_keys } => hash_keys.iter().all(|k| group_columns.contains(k)),
        }
    }
}

/// Two-valued on purpose. A whole-stream order is `BatchSorted` meeting `SingleBatch`,
/// derived by [`PartitionLayout::is_stream_sorted`] rather than declared, so there is no
/// second way to say it. It becomes a real third state only under #138's ranged merge
/// emission, which orders a stream across several batches.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum SortOrder {
    NotSpecified,
    BatchSorted { columns: Vec<ColumnOrder> },
}

impl SortOrder {
    /// An order on no columns is no order, so it canonicalizes — otherwise two layouts
    /// that mean the same thing compare unequal.
    pub fn batch_sorted(columns: Vec<ColumnOrder>) -> Self {
        if columns.is_empty() {
            Self::NotSpecified
        } else {
            Self::BatchSorted { columns }
        }
    }

    pub fn is_batch_sorted(&self) -> bool {
        matches!(self, Self::BatchSorted { .. })
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum BatchLayout {
    SingleBatch,
    MultipleBatches,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct PartitionLayout {
    pub n: usize,
    pub key_distribution: KeyDistribution,
    pub sort_order: SortOrder,
    pub batch_layout: BatchLayout,
}

impl PartitionLayout {
    /// N lanes, nothing else declared — what a scan or a shuffle-free chain emits.
    pub fn new(n: usize) -> Self {
        Self {
            n,
            key_distribution: KeyDistribution::NotSpecified,
            sort_order: SortOrder::NotSpecified,
            batch_layout: BatchLayout::MultipleBatches,
        }
    }

    /// Whole stream ordered, not merely each batch — what a top-N after a sort needs.
    pub fn is_stream_sorted(&self) -> bool {
        self.sort_order.is_batch_sorted() && self.batch_layout == BatchLayout::SingleBatch
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn key(column: u32) -> ColumnOrder {
        ColumnOrder {
            column,
            ascending: true,
            nulls_first: false,
        }
    }

    #[test]
    fn empty_sort_columns_canonicalize_to_not_specified() {
        assert_eq!(SortOrder::batch_sorted(vec![]), SortOrder::NotSpecified);
        assert!(!SortOrder::batch_sorted(vec![]).is_batch_sorted());
        assert!(SortOrder::batch_sorted(vec![key(0)]).is_batch_sorted());
    }

    #[test]
    fn stream_sorted_is_batch_sorted_meeting_single_batch() {
        let mut layout = PartitionLayout::new(1);
        assert!(!layout.is_stream_sorted());

        layout.sort_order = SortOrder::batch_sorted(vec![key(0)]);
        assert!(
            !layout.is_stream_sorted(),
            "many batches, each sorted, is not a sorted stream"
        );

        layout.batch_layout = BatchLayout::SingleBatch;
        assert!(layout.is_stream_sorted());

        layout.sort_order = SortOrder::NotSpecified;
        assert!(
            !layout.is_stream_sorted(),
            "one batch is not an ordered one"
        );
    }

    #[test]
    fn layouts_are_equal_exactly_when_every_field_agrees() {
        let base = PartitionLayout {
            n: 4,
            key_distribution: KeyDistribution::ByHash {
                hash_keys: vec![0, 2],
            },
            sort_order: SortOrder::batch_sorted(vec![key(1)]),
            batch_layout: BatchLayout::SingleBatch,
        };
        assert_eq!(base, base.clone());

        let mut lanes = base.clone();
        lanes.n = 8;
        assert_ne!(base, lanes);

        let mut keys = base.clone();
        keys.key_distribution = KeyDistribution::ByHash {
            hash_keys: vec![2, 0],
        };
        assert_ne!(base, keys, "hash key order is the routing, not a set");

        let mut sorted = base.clone();
        sorted.sort_order = SortOrder::batch_sorted(vec![ColumnOrder {
            nulls_first: true,
            ..key(1)
        }]);
        assert_ne!(base, sorted);
    }

    #[test]
    fn hash_keys_must_be_a_subset_of_the_group_columns() {
        let hashed = KeyDistribution::ByHash {
            hash_keys: vec![1, 3],
        };
        assert!(hashed.is_subset_of(&[0, 1, 3]));
        assert!(!hashed.is_subset_of(&[1, 2]));
        assert!(!KeyDistribution::NotSpecified.is_subset_of(&[0, 1]));
    }
}
